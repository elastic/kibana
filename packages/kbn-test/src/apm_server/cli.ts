/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import Readline from 'readline';

import { RunWithCommands, kibanaPackageJson, createFlagError, ToolingLog } from '@kbn/dev-utils';

import { ApmServer } from './apm_server';
import { ApmServerConfig } from './apm_server_install';
import { SnapshotBuild } from './snapshot_build';
import { Staging } from './staging';

async function userHitEnter(log: ToolingLog, msg: string) {
  log.warning(msg);
  await new Promise<void>((resolve) => {
    const rl = Readline.createInterface({
      input: process.stdin,
      output: process.stdout,
      prompt: '?',
    });

    rl.once('line', () => {
      rl.close();
      resolve();
    });
  });
}

function parseEsUrl(url: string) {
  try {
    return esUrlToConfig(new URL(url));
  } catch (error) {
    throw createFlagError(`invalid --es-url [${url}]: ${error.message}`);
  }
}

function esUrlToConfig(url: URL): NonNullable<ApmServerConfig['elasticsearch']> {
  const username = url.username;
  const password = url.password;
  url.username = '';
  url.password = '';
  return {
    hosts: [url.href],
    username,
    password,
  };
}

export function runApmServerCli() {
  new RunWithCommands({
    description: 'CLI for interacting with APM server snapshots used for Kibana development and CI',
  })
    .command({
      name: 'run',
      description: `
        Run a cached APM server artifact, or download one to run if the cached one is outdated
      `,
      flags: {
        string: ['branch', 'port', 'es-url', 'name'],
        boolean: ['staging'],
        help: `
          --branch       Specify which branch to run, current options are 'master' and '7.x', defaults to {pkg.branch}
          --staged       Run the staging artifacts that were previously downloaded using the 'staging-download' command
          --port         Port to run the apm-server on, defaults to 8200
          --es-url       Elasticsearch URL for the apm-server to index into, include auth in URL if necessary, defaults
                           to http://elastic:changeme@localhost:9200
          --name         When running multiple instances each needs a unique name, defaults to "apm-server"
        `,
      },
      async run({ log, flags }) {
        const branch = flags.branch || kibanaPackageJson.branch;
        const staging = !!flags.staging;
        const name = `${flags.name}` || undefined;

        const port = Number.parseInt(
          typeof flags.port === 'string' && flags.port ? flags.port : '8200',
          10
        );
        if (Number.isNaN(port)) {
          throw createFlagError('expected --port to be a number');
        }

        const esConfig =
          typeof flags['es-url'] === 'string' && flags['es-url']
            ? parseEsUrl(flags['es-url'])
            : undefined;

        await new ApmServer(log).run({
          name,
          branch,
          staging,
          config: {
            port,
            elasticsearch: esConfig,
          },
        });
      },
    })
    .command({
      name: 'staging-download',
      description: `
        Download artifacts from the most recent snapshot build of a specific version and stage it locally for running/testing/promoting.

        Use \`node scripts/apm_server run --staged\` to run the staged downloads.
      `,
      flags: {
        string: ['version'],
        help: `
          --version      Override the version of staged assets, use full version numbers, defaults to '{pkg.version}-SNAPSHOT'
        `,
      },
      async run({ log, flags }) {
        const version = flags.version || undefined;
        if (typeof version !== 'string' && typeof version !== 'undefined') {
          throw createFlagError('only a single --version flag is supported');
        }

        log.info('fetching snapshot builds');
        log.indent(4);
        const snapshot = await SnapshotBuild.fetchMostRecentForVersion(log, version);
        log.success(
          `using packages from build [${snapshot.id}] from ${snapshot.describeTimeSinceBuilt()}`
        );
        log.indent(-4);

        log.info('downloading snapshot packages for supported platforms');
        log.indent(4);
        const staging = new Staging(log);
        await staging.downloadBuilds(snapshot);
        log.success(`downloaded packages for build [${snapshot.id}] to staging`);
        log.indent(-4);
      },
    })
    .command({
      name: 'staging-upload',
      description: `
        Upload staged artifacts to the GCS cache. Expects a propertly configured "gsutil" executable with proper permissions.
      `,
      flags: {
        string: ['branch'],
        boolean: ['force', 'promote'],
        default: {
          promote: null,
        },
        help: `
          --branch       REQUIRED, Upload the staging artifacts to be used for a specific branch
          --force        Run the upload without asking for confirmation
          --promote      After uploading the artifacts promote them to be the "latest" as well
          --no-promote   Skip promotion without asking for confirmation
        `,
      },
      async run({ log, flags }) {
        const branch = flags.branch;
        if (!branch || typeof branch !== 'string') {
          throw createFlagError('--branch is a required flag');
        }

        const force = !!flags.force;
        const promote = flags.promote === null ? null : !!flags.promote;

        const staging = new Staging(log);
        const { artifacts, buildId } = await staging.validateAndRead();
        const artifactList = artifacts.join('\n - ');

        if (force) {
          log.warning(
            `Uploading assets without confirmation [branch=${branch}] [buildId=${buildId}]:\n - ${artifactList}`
          );
        } else {
          await userHitEnter(
            log,
            `Hit ENTER to upload the following files to GCS [branch=${branch}] [buildId=${buildId}]:\n - ${artifactList}`
          );
        }
        await staging.upload(branch, buildId);

        if (promote === false) {
          return;
        }

        if (promote === null) {
          await userHitEnter(
            log,
            `Hit ENTER to promote the uploaded files so that they will be downloaded by anyone else using this tool [branch=${branch}] [buildId=${buildId}]`
          );
        }

        log.warning(`Promoting uploaded [buildId=${buildId}] to "latest" for [branch=${branch}]`);
        await staging.promote(branch);
      },
    })
    .execute();
}
