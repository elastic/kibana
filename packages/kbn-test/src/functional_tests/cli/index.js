/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * and the Server Side Public License, v 1; you may not use this file except in
 * compliance with, at your election, the Elastic License or the Server Side
 * Public License, v 1.
 */

export { runTestsCli } from './run_tests/cli';
export { processOptions as processRunTestsCliOptions } from './run_tests/args';
export { startServersCli } from './start_servers/cli';
export { processOptions as processStartServersCliOptions } from './start_servers/args';
