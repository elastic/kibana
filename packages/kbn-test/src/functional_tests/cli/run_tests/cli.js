/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { runTests } from '../../tasks';
import { runCli } from '../../lib';
import { processOptions, displayHelp } from './args';

/**
 * Run servers and tests for each config
 * Only cares about --config option. Other options
 * are passed directly to functional_test_runner, such as
 * --bail, --verbose, etc.
 * @param {string[]} defaultConfigPaths Optional paths to configs
 *                                      if no config option is passed
 */
export async function runTestsCli(defaultConfigPaths) {
  await runCli(displayHelp, async userOptions => {
    const options = processOptions(userOptions, defaultConfigPaths);
    await runTests(options);
  });
}
