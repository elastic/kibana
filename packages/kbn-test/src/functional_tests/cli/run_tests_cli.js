import dedent from 'dedent';
import getopts from 'getopts';
import { createToolingLog, pickLevelFromFlags } from '@kbn/dev-utils';
import { runTests } from '../../';

/**
 * Run servers and tests for each config
 * Only cares about --config option. Other options
 * are passed directly to functional_test_runner, such as
 * --bail, --verbose, etc.
 * @param {string[]} defaultConfigPaths  Array of paths to configs to use
 *                                       if no config option is passed
 */
export async function runTestsCli(defaultConfigPaths) {
  const { configs, help, bail, log, installDir } = processArgs(
    defaultConfigPaths
  );

  if (help) return displayHelp();

  if (!configs || configs.length === 0) {
    log.error(
      `Run Tests requires at least one path to a config. Leave blank to use defaults.`
    );
    process.exit(9);
  }

  try {
    await runTests(configs, { bail, log, installDir });
  } catch (err) {
    log.error('FATAL ERROR');
    log.error(err);
    process.exit(1);
  }
}

function processArgs(defaultConfigPaths) {
  // If no args are passed, use {}
  const options = getopts(process.argv.slice(2)) || {};

  // If --config is passed without paths, it's "true", so use default
  const configs =
    typeof options.config === 'string' || Array.isArray(options.config)
      ? [].concat(options.config)
      : defaultConfigPaths;

  const log = createToolingLog(pickLevelFromFlags(options));
  log.pipe(process.stdout);

  return {
    configs,
    log,
    help: options.help,
    bail: options.bail,
    installDir: options['kibana-install-dir'],
    rest: options._,
  };
}

function displayHelp() {
  console.log(
    dedent(`
    Run Functional Tests

    Usage:  node scripts/functional_tests [options]

    --config      Option to pass in a config
                  Can pass in multiple configs with
                  --config file1 --config file2 --config file3
    --kibana-install-dir
                  Run Kibana from an existing install directory
                  Default: run from source
    --bail        Stop the test run at the first failure
    --help        Display this menu and exit

    Log level options:

    --verbose
    --debug
    --quiet       Log errors
    --silent
    `)
  );
}
