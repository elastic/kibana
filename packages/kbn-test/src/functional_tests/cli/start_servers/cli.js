/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { startServers } from '../../tasks';
import { runCli } from '../../lib';
import { processOptions, displayHelp } from './args';

/**
 * Start servers
 * @param {string} defaultConfigPath Optional path to config
 *                                   if no config option is passed
 */
export async function startServersCli(defaultConfigPath) {
  await runCli(displayHelp, async (userOptions) => {
    const options = processOptions(userOptions, defaultConfigPath);
    await startServers(options);
  });
}
