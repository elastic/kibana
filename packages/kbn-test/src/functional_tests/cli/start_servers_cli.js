/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import chalk from 'chalk';
import dedent from 'dedent';
import getopts from 'getopts';
import { createToolingLog, pickLevelFromFlags } from '@kbn/dev-utils';
import { startServers } from '../../';

/**
 * Start servers
 * @param {string} configPath path to config
 */
export async function startServersCli(defaultConfigPath) {
  const { config, log, help, installDir } = processArgv(defaultConfigPath);

  if (help) return displayHelp();

  if (!config) {
    log.error(
      `Start Servers requires one path to a config. Leave blank to use default.`
    );
    process.exit(1);
  }

  try {
    await startServers(config, { log, installDir });
  } catch (err) {
    log.error('FATAL ERROR');
    log.error(err);
    process.exit(1);
  }
}

function processArgv(defaultConfigPath) {
  const options = getopts(process.argv.slice(2)) || {};

  if (Array.isArray(options.config)) {
    console.log(
      chalk.red(
        `Starting servers requires a single config path. Multiple were passed.`
      )
    );
    process.exit(9);
  }

  const config =
    typeof options.config === 'string' ? options.config : defaultConfigPath;

  const log = createToolingLog(pickLevelFromFlags(options));
  log.pipe(process.stdout);

  return {
    config,
    log,
    installDir: options.kibanaInstallDir,
    help: options.help,
    rest: options._,
  };
}

function displayHelp() {
  console.log(
    dedent(`
    Start Functional Test Servers

    Usage:  node scripts/functional_tests_server [options]

    --config      Option to pass in a config
    --kibana-install-dir
                  Run Kibana from an existing install directory
                  Default: run from source
    --help        Display this menu and exit

    Log level options:

    --verbose
    --debug
    --quiet       Log errors
    --silent
    `)
  );
}
