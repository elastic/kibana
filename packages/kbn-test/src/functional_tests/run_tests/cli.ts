/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { run } from '@kbn/dev-cli-runner';

import { initLogsDir } from '../lib/logs_dir';
import { runTests } from './run_tests';
import { parseFlags, FLAG_OPTIONS } from './flags';

export function runTestsCli() {
  run(
    async ({ flagsReader, log }) => {
      const options = parseFlags(flagsReader);

      if (options.logsDir) {
        initLogsDir(log, options.logsDir);
      }

      await runTests(log, options);
    },
    {
      description: `Run Functional Tests`,
      usage: `
        node scripts/functional_tests --help
        node scripts/functional_tests [--config <file1> [--config <file2> ...]]
        node scripts/functional_tests [options] [-- --<other args>]
      `,
      flags: FLAG_OPTIONS,
    }
  );
}
