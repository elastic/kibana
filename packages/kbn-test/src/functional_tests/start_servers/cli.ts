/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { run } from '@kbn/dev-cli-runner';

import { initLogsDir } from '../lib/logs_dir';

import { parseFlags, FLAG_OPTIONS } from './flags';
import { startServers } from './start_servers';

/**
 * Start servers
 */
export function startServersCli() {
  run(
    async ({ flagsReader: flags, log }) => {
      const options = parseFlags(flags);

      if (options.logsDir) {
        initLogsDir(log, options.logsDir);
      }

      await startServers(log, options);
    },
    {
      flags: FLAG_OPTIONS,
    }
  );
}
