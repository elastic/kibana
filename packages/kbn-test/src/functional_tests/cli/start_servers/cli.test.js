/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { Writable } from 'stream';

import { startServersCli } from './cli';
import { checkMockConsoleLogSnapshot } from '../../test_helpers';

// Note: Stub the startServers function to keep testing only around the cli
// method and arguments.
jest.mock('../../tasks', () => ({
  startServers: jest.fn(),
}));

describe('start servers CLI', () => {
  describe('options', () => {
    const originalObjects = { process, console };
    const exitMock = jest.fn();
    const logMock = jest.fn(); // mock logging so we don't send output to the test results
    const argvMock = ['foo', 'foo'];

    const processMock = {
      exit: exitMock,
      argv: argvMock,
      stdout: new Writable(),
      cwd: jest.fn(),
      env: {
        ...originalObjects.process.env,
        TEST_ES_FROM: 'snapshot',
      },
    };

    beforeAll(() => {
      global.process = processMock;
      global.console = { log: logMock };
    });

    afterAll(() => {
      global.process = originalObjects.process;
      global.console = originalObjects.console;
    });

    beforeEach(() => {
      global.process.argv = [...argvMock];
      global.process.env = {
        ...originalObjects.process.env,
        TEST_ES_FROM: 'snapshot',
      };
      jest.resetAllMocks();
    });

    it('rejects boolean config value', async () => {
      global.process.argv.push('--config');

      await startServersCli();

      expect(exitMock).toHaveBeenCalledWith(1);
      checkMockConsoleLogSnapshot(logMock);
    });

    it('rejects empty config value if no default passed', async () => {
      global.process.argv.push('--config', '');

      await startServersCli();

      expect(exitMock).toHaveBeenCalledWith(1);
      checkMockConsoleLogSnapshot(logMock);
    });

    it('accepts empty config value if default passed', async () => {
      global.process.argv.push('--config', '');

      await startServersCli('foo');

      expect(exitMock).not.toHaveBeenCalled();
    });

    it('rejects bail', async () => {
      global.process.argv.push('--bail', true);

      await startServersCli('foo');

      expect(exitMock).toHaveBeenCalledWith(1);
      checkMockConsoleLogSnapshot(logMock);
    });

    it('accepts string value for kibana-install-dir', async () => {
      global.process.argv.push('--kibana-install-dir', 'foo');

      await startServersCli('foo');

      expect(exitMock).not.toHaveBeenCalled();
    });

    it('rejects boolean value for kibana-install-dir', async () => {
      global.process.argv.push('--kibana-install-dir');

      await startServersCli('foo');

      expect(exitMock).toHaveBeenCalledWith(1);
      checkMockConsoleLogSnapshot(logMock);
    });

    it('accepts boolean value for updateBaselines', async () => {
      global.process.argv.push('--updateBaselines');

      await startServersCli('foo');

      expect(exitMock).toHaveBeenCalledWith(1);
      checkMockConsoleLogSnapshot(logMock);
    });

    it('accepts boolean value for updateSnapshots', async () => {
      global.process.argv.push('--updateSnapshots');

      await startServersCli('foo');

      expect(exitMock).toHaveBeenCalledWith(1);
      checkMockConsoleLogSnapshot(logMock);
    });

    it('accepts source value for esFrom', async () => {
      global.process.argv.push('--esFrom', 'source');

      await startServersCli('foo');

      expect(exitMock).not.toHaveBeenCalled();
    });

    it('accepts debug option', async () => {
      global.process.argv.push('--debug');

      await startServersCli('foo');

      expect(exitMock).not.toHaveBeenCalled();
    });

    it('accepts silent option', async () => {
      global.process.argv.push('--silent');

      await startServersCli('foo');

      expect(exitMock).not.toHaveBeenCalled();
    });

    it('accepts quiet option', async () => {
      global.process.argv.push('--quiet');

      await startServersCli('foo');

      expect(exitMock).not.toHaveBeenCalled();
    });

    it('accepts verbose option', async () => {
      global.process.argv.push('--verbose');

      await startServersCli('foo');

      expect(exitMock).not.toHaveBeenCalled();
    });

    it('accepts extra server options', async () => {
      global.process.argv.push('--', '--server.foo=bar');

      await startServersCli('foo');

      expect(exitMock).not.toHaveBeenCalled();
    });

    it('accepts help option even if invalid options passed', async () => {
      global.process.argv.push('--debug', '--grep', '--help');

      await startServersCli('foo');

      expect(exitMock).not.toHaveBeenCalledWith(1);
    });

    it('rejects invalid options even if valid options exist', async () => {
      global.process.argv.push('--debug', '--grep', '--bail');

      await startServersCli('foo');

      expect(exitMock).toHaveBeenCalledWith(1);
      checkMockConsoleLogSnapshot(logMock);
    });
  });
});
