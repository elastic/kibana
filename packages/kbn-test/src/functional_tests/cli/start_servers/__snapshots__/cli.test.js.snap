// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`start servers CLI options accepts boolean value for updateBaselines 1`] = `
"
[31mfunctional_tests_server: invalid option [updateBaselines][39m
    ...stack trace...
"
`;

exports[`start servers CLI options rejects bail 1`] = `
"
[31mfunctional_tests_server: invalid option [bail][39m
    ...stack trace...
"
`;

exports[`start servers CLI options rejects boolean config value 1`] = `
"
[31mfunctional_tests_server: invalid argument [true] to option [config][39m
    ...stack trace...
"
`;

exports[`start servers CLI options rejects boolean value for kibana-install-dir 1`] = `
"
[31mfunctional_tests_server: invalid argument [true] to option [kibana-install-dir][39m
    ...stack trace...
"
`;

exports[`start servers CLI options rejects empty config value if no default passed 1`] = `
"
[31mfunctional_tests_server: config is required[39m
    ...stack trace...
"
`;

exports[`start servers CLI options rejects invalid options even if valid options exist 1`] = `
"
[31mfunctional_tests_server: invalid option [grep][39m
    ...stack trace...
"
`;

exports[`start servers CLI options rejects non-enum value for esFrom 1`] = `
"
[31mfunctional_tests_server: invalid argument [butter] to option [esFrom][39m
    ...stack trace...
"
`;
